app/helpers/ExerciseConfig/Helper.php

```diff
@@ -93,10 +134,119 @@ public function getVariablesForExercise(array $pipelines, VariablesTable $enviro
+    $envStatuses = [];
+
+    $envConfigs = [];
+    foreach ($exercise->getRuntimeEnvironments() as $environment) {
+      $exerciseEnvironment = $exercise->getExerciseEnvironmentConfigByEnvironment($environment);
+      if ($exerciseEnvironment === null) {
+        throw new ExerciseConfigException("Environment config not found for environment '{$environment->getId()}'");
+      }
+
+      $parsedConfig = $exerciseEnvironment->getParsedVariablesTable();
+      $envConfig = $this->loader->loadVariablesTable($parsedConfig);
+      $envConfigs[$environment->getId()] = $envConfig;
+      $envStatuses[$environment->getId()] = true;
+    }
+
+    $config = $this->loader->loadExerciseConfig($exercise->getExerciseConfig()->getParsedConfig());
+    foreach ($config->getTests() as $testId => $test) {
+      foreach ($test->getEnvironments() as $envId => $env) {
+        $envConfig = $envConfigs[$envId];
+
+        // load all pipelines for this test and environment
+        $pipelines = [];
+        foreach ($env->getPipelines() as $pipelineId => $pipeline) {
+          $pipelineEntity = $this->pipelines->findOrThrow($pipelineId);
+          $pipelineConfig = $this->loader->loadPipeline($pipelineEntity->getPipelineConfig()->getParsedPipeline());
+          $pipelines[$pipelineId] = $pipelineConfig;
+        }
+
+        // join pipelines and return inputs from all of them
+        $inputs = []; // pairs of pipeline identifier and port indexed by variable name
+        $references = []; // pairs of pipeline identifier and variable indexed by variable name
+        $this->joinPipelinesAndGetInputVariables($pipelines, $inputs, $references);
+
+        // go through all inputs, references are no use to us
+        foreach ($inputs as $varName => $input) {
+          if (!$input[1]->isFile()) {
+            continue;
+          }
+
+          // now we have only file input ports... use them... for the glory of ReCodEx
+
+          // try to find variable in environment config variable table
+          $variable = $envConfig->get($varName);
+          if ($variable === null) {
+            // if variable was not found in environment config, peek into exercise config
+            $variable = $env->getPipeline($input[0])->getVariablesTable()->get($varName);
+          }
+
+          // somethings fishy here
+          if ($variable === null) {
+            throw new ExerciseConfigException("Variable '{$varName}' not found in environment and exercise config");
+          }
+
+          // we only seek for the local file variables, not the remote ones...
+          // although port might be of type file, variables assigned to it
+          // may have remote-file type in case the file should be downloaded
+          if (!$variable->isFile()) {
+            continue;
+          }
+
+          // determine if variable really defines only single wildcard expression
+          // and not an array of values
+          if ($variable->isValueArray()) {
```

Comment:
This may not be entirely correct. What if someone explicitly requires files with names 'implementation.hpp' and 'implementation.cpp'?